include/rosidl_dynamic_typesupport/uchar.h

```diff
@@ -26,7 +26,7 @@ extern "C" {
 #    define INCLUDED_UCHAR 1
 #  endif
 #endif
-#if !defined(INCLUDED_UCHAR)
+#if !defined(INCLUDED_UCHAR) && __cplusplus <= 199711L
```

Comment:
This will only compile if  `__cplusplus` is  defined, no?

Reply:
I guess that we shall also add a check on defined(__cplusplus). Can you take a look @Acuadros95?

Reply:
Undefined values on macros are replaced with a `0` value, so if `__cplusplus` is not defined this condition will always be true: `__cplusplus <= 199711L`.

Reply:
As @clalancette mentioned, this does not make sense. We use C++14/C++17 with gcc and msvc and this compiles fine. According to the standard, this type is defined only when `<uchar.h>` is included:

https://en.cppreference.com/w/c/string/multibyte/char16_t

Since this header is not available on some machines, e.g. macOS and Cygwin, we added this bit of code. I think we need some checks to avoid making this type def in more cases (like yours), but it's not because of the C++ version. It likely has to do with the std library you are using in microROS and that is defines this type (`char16_t`) in a different header that is being included automatically here, e.g. maybe via `stdint.h` or something else in the `.c` file that's failing to compile.

Reply:
> We use C++14/C++17 with gcc and msvc and this compiles fine. According to the standard, this type is defined only when <uchar.h> is included:

That is for C, for C++11 and beyond its a keyword and cannot be redefined: 

https://en.cppreference.com/w/cpp/keyword/char16_t

I guess the problem is that our compiler does not have the `__has_include` definition, and the `<uchar.h>` header is most probably checking the used C++ version to add the typedef or not, example:

https://codebrowser.dev/llvm/include/uchar.h.html#34

Reply:
> I guess the problem is that our compiler does not have the `__has_include` definition, and the `<uchar.h>` header is most probably checking the used C++ version to add the typedef or not, example:

I see.  That makes a whole lot more sense.  Thanks.  In that case, then I think we should probably rewrite this whole section (including the stuff above) to something like:

```
#if defined(__cplusplus) && __cplusplus >= 201103L
// Nothing to do here, C++11 and beyond have char16_t as a keyword:
// https://en.cppreference.com/w/cpp/keyword/char16_t
#elif defined(__has_include) && __has_include(<uchar.h>)
// If the compiler has __has_include, and uchar.h exists, include that as it will have char16_t
// as a typedef.
#  include <uchar.h>
#else
// Otherwise assume that char16_t isn't defined anywhere, and define it ourselves as uint_least16_t.
#  include <stdint.h>
typedef uint_least16_t char16_t;
#endif
```

That is, if we are on C++11 or newer, just assume it is a keyword.  If we aren't on C++, or it is too old, then try to use `__has_include` to find the `uchar.h` header and include that.  If all of that fails, assume `char16_t` isn't defined and define it ourselves.  @pablogs9 @Acuadros95 can you give this a try and see if it works with your compiler?

Reply:
Sounds good. @Acuadros95 can you try it?

Reply:
@clalancette Just tested the new checks and can confirm it works.

I have updated the PR here https://github.com/ros2/rosidl_dynamic_typesupport/pull/3/commits/ac34007e3018eee651e994874680c8043c632a3e